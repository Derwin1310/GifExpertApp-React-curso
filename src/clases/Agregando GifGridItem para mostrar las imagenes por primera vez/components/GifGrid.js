import React, { useEffect, useState } from 'react'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    //Aqui usamos el useState de esta funcion para poder imprimirla mas abajo con los gif en el li, y lo hacemos un array vacio para que tenga donde almacenar las imagenes
    const [images, setImages] = useState([])

    useEffect(() => {
        getGifs();
    }, [])

    const getGifs = async () => {
        const url = 'https://api.giphy.com/v1/gifs/search?q=Rick and Morty&limit=12&api_key=46Sat15Rc1PEzzKY0bFdQs2ly07cuhtk';
        const res = await fetch(url);
        const { data } = await res.json();

        const gif = data.map(img => {
            return {
                id: img.id,
                title: img.title,
                url: img.images?.downsized_medium.url,
            };
        });
        console.log(gif)
        //Aqui es donde le colocamos los gif al setImages para que al momento de ejecutar la funcion se muestren las imagenes
        setImages(gif)
    };


    return (
        <div>
            <h3>{category}</h3>
            <ol>
                {
                    images.map(img => (
                        <GifGridItem
                            key={img.id}
                            //Aqui le asignamos el operador spread a img para que se traiga todo lo que recorre el images.map que son basicamente las imagenes de getGifs y no tener que colocar el url, title y esas cosas de manera manual, en pocas palabras le estamos asignando como parametro los id, titles y url de las imagenes a GGI.
                            {...img}
                        />
                    ))
                }
            </ol>
        </div>
    )
}
