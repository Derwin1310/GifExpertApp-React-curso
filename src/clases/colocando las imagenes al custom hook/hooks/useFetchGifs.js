import {useState, useEffect} from 'react';
import {getGifs} from '../helpers/GetGifs';

//Aqui estamos realizando el custom hook como una funcion comun y corriente, solo que al colocarle la palabra use al inicio ya lo podemos manejar como un hook, y lo que hace es una funcion sencilla que retorna el nuevo valor del state tras un setTimeOut para que aplique la logica en el GifGrid.
export const useFetchGifs = category => {
	const [state, setState] = useState({
		data: [],
		loading: true,
	});

    //Podemos aplicar el useEffect asi como el useState dentro de nuestros custom hooks para que lo renderize una sola vez.
	useEffect(() => {
        //Recordemos que este category es el parametro que trae desde el GifGrid que viene de GifExpertApp para mostrar las imagenes
		getGifs(category).then(img => {
            //El setTimeOut es para simular una carga mas lenta de la api
			setTimeout(() => {
                console.log(img)
                //Y usamos la funcion del useState para resolver la promesa del getGifs volviendo la data: img para destructurarla en GifGrid
				setState({
					data: img,
					loading: false,
				});
			}, 3000);
		});
	}, [category]);

	return state;
};
