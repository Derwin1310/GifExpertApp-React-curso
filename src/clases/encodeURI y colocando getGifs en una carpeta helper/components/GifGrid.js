import React, {useEffect, useState} from 'react';
import {getGifs} from '../helpers/GetGifs';
import {GifGridItem} from './GifGridItem';

export const GifGrid = ({category}) => {
	//Aqui usamos el useState de esta funcion para poder imprimirla mas abajo con los gif en el li, y lo hacemos un array vacio para que tenga donde almacenar las imagenes
	const [images, setImages] = useState([]);

	useEffect(() => {
		//Lo que hicimos aqui fue borrar el componente que teniamos de getGifs y lo pasamos a otro archivo, de esa manera solo estamos trayendo la funcion para ejecutarla aqui y no tener tanto codigo en este archivo.
		//Y le colocamos el .then debido a que retorna una promesa al ser una funcion async
		getGifs(category)
            .then(setImages);
	}, [category]);

	return (
		<>
			<h3>{category}</h3>
			<div className='card-grid'>
				<ol>
					{images.map(img => (
						<GifGridItem
							key={img.id}
							//Aqui le asignamos el operador spread a img para que se traiga todo lo que recorre el images.map que son basicamente las imagenes de getGifs y no tener que colocar el url, title y esas cosas de manera manual, en pocas palabras le estamos asignando como parametro los id, titles y url de las imagenes a GGI.
							{...img}
						/>
					))}
				</ol>
			</div>
		</>
	);
};
