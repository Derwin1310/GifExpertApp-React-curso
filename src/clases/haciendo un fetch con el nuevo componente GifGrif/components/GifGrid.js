import React from 'react'

//Aqui estamos creando el componente funcional y exportandolo para usarlo en el GEA
export const GifGrid = ({ category }) => {

    //Como estamos haciendo una peticion recordamos que tenemos que hacerla asincrona.
    const getGifs = async () => {
        const url = 'https://api.giphy.com/v1/gifs/search?q=Rick and Morty&limit=10&api_key=46Sat15Rc1PEzzKY0bFdQs2ly07cuhtk';
        const res = await fetch(url);
        const {data} = await res.json();

        //aqui creamos esta constante  llamada gif para que muestra la data desestructurada de la respuesta del formato json que nos llega de la peticion fetch
        //Usamos el .map para que recorra toda la data y nos devuelva un array nuevo con los elementos que le estamos pidiendo
        const gif = data.map(img => {
            return {
                id: img.id,
                title: img.title,
                url: img.images?.downsized_medium.url,
            };
        });
        console.log(gif)
    };

    getGifs();

    return (
        <div>
            {/* Este category es el que estamos mostrando en pantalla del input de AddCategory */}
            <h3>{category}</h3>
        </div>
    )
}
