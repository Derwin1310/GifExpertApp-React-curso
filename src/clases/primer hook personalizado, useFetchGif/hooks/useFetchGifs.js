import {useState} from 'react';

//Aqui estamos realizando el custom hook como una funcion comun y corriente, solo que al colocarle la palabra use al inicio ya lo podemos manejar como un hook, y lo que hace es una funcion sencilla que retorna el nuevo valor del state tras un setTimeOut para que aplique la logica en el GifGrid.
export const useFetchGifs = () => {
	const [state, setState] = useState({
		data: [],
		loading: true,
	});

    setTimeout(() => {
        setState({
            data: [1,2,3],
            loading: false,
        })
    },3000);

    return state;
};
