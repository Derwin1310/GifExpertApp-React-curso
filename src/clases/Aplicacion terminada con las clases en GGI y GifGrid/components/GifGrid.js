import React from 'react';
import {useFetchGifs} from '../hooks/useFetchGifs';
import {GifGridItem} from './GifGridItem';

export const GifGrid = ({category}) => {
	const {data: imgs, loading} = useFetchGifs(category);

	return (
		<>
            {/* Aqui mostramos los estilos al h3 que muestra lo que tiene category cuando insertamos algo en el input y al loading mientras cargan con el setTimeOut.*/}
			<h3 className='card animate__animated animate__bounceInRight animate__delay-5s'>
				{category}
			</h3>
			{loading && (
				<h4 className='animate__animated animate__flash'>Loading...</h4>
			)}
			<div className='card-grid'>
				{imgs.map(img => (
					<GifGridItem key={img.id} {...img} />
				))}
			</div>
		</>
	);
};
