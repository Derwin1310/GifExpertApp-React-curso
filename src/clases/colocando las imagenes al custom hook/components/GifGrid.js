import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import {GifGridItem} from './GifGridItem';

export const GifGrid = ({category}) => {
	
    //Aqui lo que hacemos es aplicar el custom hook pero ya utilizandolo de verdad con la peticion https del fetch y le colocamos el category para que muestre las imagenes.
    //De modo que ya no necesitamos el getGifs
    const {data: imgs, loading} = useFetchGifs(category);
    
	return (
		<>
			<h3>{category}</h3>
            {/* Aqui colocamos el loading con una condicion para que muestre un parrafo mientras vienen las imagenes */}
            {loading && <h4>Loading...</h4>}
			<div className='card-grid'>
                {/* Y aqui le hacemos el .map al imgs ya que el es quien tiene la funcion del getGifs para traer las imagenes */}
				{imgs.map(img => (
					<GifGridItem key={img.id} {...img} />
				))}
			</div>
		</>
	);
};
