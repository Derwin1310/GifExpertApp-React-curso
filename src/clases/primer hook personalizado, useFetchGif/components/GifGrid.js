import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
// import {getGifs} from '../helpers/GetGifs';
// import {GifGridItem} from './GifGridItem';

export const GifGrid = ({category}) => {
	/* const [images, setImages] = useState([]);
    

	useEffect(() => {
		getGifs(category).then(setImages);
	}, [category]); */

    //En esta parte reemplazamos el codigo de arriba con este custom hook que nos permite cargar la informacion que le enviemos, el archivo se encuentra en hooks.
    //Aqui lo que hacemos es destructurar el objeto que tiene useFetchGifs()
    const {loading} = useFetchGifs();
    
	return (
		<>
			<h3>{category}</h3>
            {/*Y colocamos en esta parte el loading para que se ejecute la funcion del custom hook */}
            {loading ? 'Loading...' : 'Data succesfully charged'}
			{/* <div className='card-grid'>
				{images.map(img => (
					<GifGridItem key={img.id} {...img} />
				))}
			</div> */}
		</>
	);
};
