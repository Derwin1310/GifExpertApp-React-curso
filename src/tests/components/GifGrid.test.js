import {GifGrid} from '../../components/GifGrid';
import {shallow} from 'enzyme';
import {useFetchGifs} from '../../hooks/useFetchGifs';
//Este jest.mock con el path de nuestro useFetchGifs hace que podamos utilizar este custom hook en los test para diferentes pruebas y no tener que importar las dependencias que este contiene.
jest.mock('../../hooks/useFetchGifs');

describe('Tests in <GifGrid />', () => {
	const category = 'Attack on Titan';

    //En este test hacemos un snapshot de como queda el jsx antes de usar al custom hook
	test('should show <GifGrid /> correctly', () => {
        //usamos el useFetchGifs con el return value para simular que lo estamos usando con una data que no existe en realidad.
        //En este test lo probamos como si no se usara.
		useFetchGifs.mockReturnValue({
			data: [],
			loading: true,
		});

		const wrapper = shallow(<GifGrid category={category} />);
		expect(wrapper).toMatchSnapshot();
	});

    //Y en este mostramos como queda el jsx despues de usar el custom hook.
	test('should show items when useFetchGifs get used', () => {
        //Aqui creamos un array con un objeto para simular que esto es lo que estamos devolviendo en el custom hook
		const gifs = [
			{
				id: 'Derwin',
				title: 'Im a title',
				url: 'https://localhost/any/anything.jpg',
			},
			{
				id: 'Eduardo',
				title: 'Im a title vol.2',
				url: 'https://localhost/any/anythingVol2.jpg',
			},
		];

        //Y una vez este se crea el custom hook ahora le colocamos los valores que deberia tener cuando la data cambia por un objeto que se trajo y hacemos un snapshot para mostrar como queda.
		useFetchGifs.mockReturnValue({
			data: gifs,
			loading: false,
		});

		const wrapper = shallow(<GifGrid category={category} />);

		expect(wrapper).toMatchSnapshot();
		//Aqui evaluanos sobre si el parrafo que dice Loading no existe cuando esta en false.
		expect(wrapper.find('p').exists()).toBe(false);
		//Aqui evaluamos si el GGI que es quien recibe el array de objetos de gifs es igual a gifs
		expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
		//Aqui mas de lo mismo con lo de arriba(importante no redundar, solo lo hice por temas de aprendizaje )
		expect(wrapper.find('GifGridItem').length).toBe(2);
	});
});
