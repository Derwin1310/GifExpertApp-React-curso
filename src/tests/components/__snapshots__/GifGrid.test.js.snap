// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <GifGrid /> should show <GifGrid /> correctly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__bounceInRight animate__delay-5s"
  >
    Attack on Titan
  </h3>
  <h4
    className="animate__animated animate__flash"
  >
    Loading...
  </h4>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Tests in <GifGrid /> should show items when useFetchGifs get used 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__bounceInRight animate__delay-5s"
  >
    Attack on Titan
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="Derwin"
      key="Derwin"
      title="Im a title"
      url="https://localhost/any/anything.jpg"
    />
    <GifGridItem
      id="Eduardo"
      key="Eduardo"
      title="Im a title vol.2"
      url="https://localhost/any/anythingVol2.jpg"
    />
  </div>
</Fragment>
`;
